BFGS/bfgs.cpp

```diff
@@ -0,0 +1,102 @@
+// Test BFGS method on problems (serial)
+#include <iostream>
+#include <cmath>
+#include <fstream>
+#include "../test/testfun.hpp"
+#include <Eigen/Dense>
+
+using namespace std;
+
+template<typename ObjFunc> // Use the parameter 'max_bound' to define edge of hypercube to optimse in
+Eigen::VectorXd bfgs(ObjFunc f, const int max_iter, const Eigen::VectorXd& int_val,
+  const double rel_sol_change_tol, const double grad_nrm_tol, const double gam,
+   const double max_bound){
+    // Initialise vectors to hold xk, xk+1 and print initial value.
+    Eigen::VectorXd step_prev = int_val, step;
+
+    // Initialise dimension and gradient vectors of xk, xk+1 respectively
+    const int dim = int_val.size();
+    Eigen::VectorXd df_prev(dim), df(dim), s(dim), q(dim), w(dim);
+
+    // Useful constants for updates
+    double st_q, qt_d_q;
+
+    // Initialise relative changes to large number
+    double rel_sol_change = 2*rel_sol_change_tol;
+    double grad_nrm = 2*grad_nrm_tol;
+
+    // Initialise matrix D as the identity matrix
+    Eigen::MatrixXd d = Eigen::MatrixXd::Identity(dim, dim);
+
+    int iter = 0;  // Initial no. iterations
+
+    // Compute initial gradient
+    df_prev = f.gradient(step_prev);
+
+    // While max iteraitons has not been reached and all change tolerances have
+    // not been met, perform bfgs update
+    while (iter < max_iter && rel_sol_change > rel_sol_change_tol && grad_nrm > grad_nrm_tol){
+      // Update step
+      step = step_prev - gam*d*df_prev;
+
+      // Project iteration back into hypercube if needed
+      for (int i=0; i<dim; ++i){
+        if (step(i) < -max_bound){step(i) = -max_bound;}
+        else if (step(i)>max_bound){step(i)=max_bound;}
+      }
+
```

Comment:
Nice! Simple and works. We'll have to add a min bound for the Shekel function later.